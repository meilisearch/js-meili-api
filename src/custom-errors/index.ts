import {
  AxiosError,
} from 'axios'
import MeiliSearchApiError from './meilisearch-error'
import MeilISearchCommunicationError from './meilisearch-communication-error'

function errorHandler (e: AxiosError, cachedStack?: string): void {
    if (e.response) {
      throw new MeiliSearchApiError(e, cachedStack)
    }
    else if (e.isAxiosError) {
      throw new MeilISearchCommunicationError(e.message)
    }
    else {
      throw e;
    }
  }

export { errorHandler }
