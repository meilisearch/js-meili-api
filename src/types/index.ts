export * from "./settings.js";
export * from "./types.js";
