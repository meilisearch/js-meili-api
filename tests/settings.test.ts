import { ErrorStatusCode, Settings } from '../src/types'
import {
  clearAllIndexes,
  config,
  BAD_HOST,
  MeiliSearch,
  getClient,
  dataset,
} from './utils/meilisearch-test-utils'

const index = {
  uid: 'movies_test',
}
const indexAndPK = {
  uid: 'movies_test_with_pk',
  primaryKey: 'id',
}

jest.setTimeout(100 * 1000)

afterAll(() => {
  return clearAllIndexes(config)
})

describe.each([{ permission: 'Master' }, { permission: 'Admin' }])(
  'Test on settings',
  ({ permission }) => {
    beforeEach(async () => {
      await clearAllIndexes(config)
      const client = await getClient('Master')
      const { taskUid: AddDocPkTask } = await client
        .index(indexAndPK.uid)
        .addDocuments(dataset, {
          primaryKey: indexAndPK.primaryKey,
        })
      await client.waitForTask(AddDocPkTask)

      const { taskUid: AddDocTask } = await client
        .index(index.uid)
        .addDocuments(dataset, {})
      await client.waitForTask(AddDocTask)
    })

    test(`${permission} key: Get default settings of an index`, async () => {
      const client = await getClient(permission)

      const response = await client.index(index.uid).getSettings()

      expect(response).toMatchSnapshot()
    })

    test(`${permission} key: Get default settings of empty index with primary key`, async () => {
      const client = await getClient(permission)

      const response = await client.index(indexAndPK.uid).getSettings()

      expect(response).toMatchSnapshot()
    })

    test(`${permission} key: Update settings`, async () => {
      const client = await getClient(permission)
      const newSettings: Settings = {
        filterableAttributes: ['title'],
        sortableAttributes: ['title'],
        distinctAttribute: 'title',
        searchableAttributes: ['title'],
        displayedAttributes: ['title'],
        rankingRules: ['id:asc', 'typo'],
        stopWords: ['the'],
        synonyms: { harry: ['potter'] },
        typoTolerance: {
          enabled: false,
          minWordSizeForTypos: {
            oneTypo: 1,
            twoTypos: 100,
          },
          disableOnWords: ['prince'],
          disableOnAttributes: ['comment'],
        },
        pagination: {
          maxTotalHits: 1000,
        },
        faceting: {
          maxValuesPerFacet: 50,
          sortFacetValuesBy: {
            '*': 'alpha',
          },
        },
        separatorTokens: ['&sep', '/', '|'],
        nonSeparatorTokens: ['&sep', '/', '|'],
        dictionary: ['J. K.', 'J. R. R.'],
      }
      // Add the settings
      const task = await client.index(index.uid).updateSettings(newSettings)
      await client.index(index.uid).waitForTask(task.taskUid)

      // Fetch the settings
      const response = await client.index(index.uid).getSettings()

      // tests
      expect(response).toMatchSnapshot()
    })

    test(`${permission} key: Update settings with all null values`, async () => {
      const client = await getClient(permission)
      const newSettings = {
        filterableAttributes: null,
        sortableAttributes: null,
        distinctAttribute: null,
        searchableAttributes: null,
        displayedAttributes: null,
        rankingRules: null,
        stopWords: null,
        synonyms: null,
        typoTolerance: {
          enabled: null,
          minWordSizeForTypos: {
            oneTypo: null,
            twoTypos: null,
          },
          disableOnWords: null,
          disableOnAttributes: null,
        },
        faceting: {
          maxValuesPerFacet: null,
          sortFacetValuesBy: null,
        },
        pagination: {
          maxTotalHits: null,
        },
        separatorTokens: null,
        nonSeparatorTokens: null,
        dictionary: null,
      }
      // Add the settings
      const task = await client.index(index.uid).updateSettings(newSettings)
      await client.index(index.uid).waitForTask(task.taskUid)

      // Fetch the settings
      const response = await client.index(index.uid).getSettings()

      // tests
      expect(response).toMatchSnapshot()
    })

    test(`${permission} key: Update settings on empty index with primary key`, async () => {
      const client = await getClient(permission)
      const newSettings = {
        distinctAttribute: 'title',
        rankingRules: ['title:asc', 'typo'],
        stopWords: ['the'],
      }
      const task = await client
        .index(indexAndPK.uid)
        .updateSettings(newSettings)
      await client.index(indexAndPK.uid).waitForTask(task.taskUid)

      const response = await client.index(indexAndPK.uid).getSettings()

      expect(response).toMatchSnapshot()
    })

    test(`${permission} key: Reset settings`, async () => {
      const client = await getClient(permission)
      const task = await client.index(index.uid).resetSettings()
      await client.index(index.uid).waitForTask(task.taskUid)

      const response = await client.index(index.uid).getSettings()

      expect(response).toMatchSnapshot()
    })

    test(`${permission} key: Reset settings of empty index`, async () => {
      const client = await getClient(permission)
      const task = await client.index(indexAndPK.uid).resetSettings()
      await client.index(index.uid).waitForTask(task.taskUid)

      const response = await client.index(indexAndPK.uid).getSettings()

      expect(response).toMatchSnapshot()
    })

    test(`${permission} key: Update searchableAttributes settings on empty index`, async () => {
      const client = await getClient(permission)
      const newSettings = {
        searchableAttributes: ['title'],
      }
      const task = await client.index(index.uid).updateSettings(newSettings)
      await client.index(index.uid).waitForTask(task.taskUid)

      const response = await client.index(index.uid).getSettings()

      expect(response).toMatchSnapshot()
    })

    test(`${permission} key: Update searchableAttributes settings on empty index with a primary key`, async () => {
      const client = await getClient(permission)
      const newSettings = {
        searchableAttributes: ['title'],
      }
      // Update settings
      const task = await client
        .index(indexAndPK.uid)
        .updateSettings(newSettings)
      // Wait for setting addition to be done
      await client.index(index.uid).waitForTask(task.taskUid)

      // Fetch settings
      const response = await client.index(indexAndPK.uid).getSettings()

      expect(response).toMatchSnapshot()
    })
  }
)

describe.each([{ permission: 'Search' }])(
  'Test on settings',
  ({ permission }) => {
    beforeEach(async () => {
      await clearAllIndexes(config)
    })
    test(`${permission} key: try to get settings and be denied`, async () => {
      const client = await getClient(permission)
      await expect(
        client.index(index.uid).getSettings()
      ).rejects.toHaveProperty('code', ErrorStatusCode.INVALID_API_KEY)
    })
    test(`${permission} key: try to update settings and be denied`, async () => {
      const client = await getClient(permission)
      await expect(
        client.index(index.uid).updateSettings({})
      ).rejects.toHaveProperty('code', ErrorStatusCode.INVALID_API_KEY)
    })
    test(`${permission} key: try to reset settings and be denied`, async () => {
      const client = await getClient(permission)
      await expect(
        client.index(index.uid).resetSettings()
      ).rejects.toHaveProperty('code', ErrorStatusCode.INVALID_API_KEY)
    })
  }
)

describe.each([{ permission: 'No' }])('Test on settings', ({ permission }) => {
  beforeEach(async () => {
    await clearAllIndexes(config)
  })
  test(`${permission} key: try to get settings and be denied`, async () => {
    const client = await getClient(permission)
    await expect(client.index(index.uid).getSettings()).rejects.toHaveProperty(
      'code',
      ErrorStatusCode.MISSING_AUTHORIZATION_HEADER
    )
  })
  test(`${permission} key: try to update settings and be denied`, async () => {
    const client = await getClient(permission)
    await expect(
      client.index(index.uid).updateSettings({})
    ).rejects.toHaveProperty(
      'code',
      ErrorStatusCode.MISSING_AUTHORIZATION_HEADER
    )
  })
  test(`${permission} key: try to reset settings and be denied`, async () => {
    const client = await getClient(permission)
    await expect(
      client.index(index.uid).resetSettings()
    ).rejects.toHaveProperty(
      'code',
      ErrorStatusCode.MISSING_AUTHORIZATION_HEADER
    )
  })
})

describe.each([
  { host: BAD_HOST, trailing: false },
  { host: `${BAD_HOST}/api`, trailing: false },
  { host: `${BAD_HOST}/trailing/`, trailing: true },
])('Tests on url construction', ({ host, trailing }) => {
  test(`Test getSettings route`, async () => {
    const route = `indexes/${index.uid}/settings`
    const client = new MeiliSearch({ host })
    const strippedHost = trailing ? host.slice(0, -1) : host
    await expect(client.index(index.uid).getSettings()).rejects.toHaveProperty(
      'message',
      `request to ${strippedHost}/${route} failed, reason: connect ECONNREFUSED ${BAD_HOST.replace(
        'http://',
        ''
      )}`
    )
  })

  test(`Test updateSettings route`, async () => {
    const route = `indexes/${index.uid}/settings`
    const client = new MeiliSearch({ host })
    const strippedHost = trailing ? host.slice(0, -1) : host
    await expect(
      client.index(index.uid).updateSettings({})
    ).rejects.toHaveProperty(
      'message',
      `request to ${strippedHost}/${route} failed, reason: connect ECONNREFUSED ${BAD_HOST.replace(
        'http://',
        ''
      )}`
    )
  })

  test(`Test resetSettings route`, async () => {
    const route = `indexes/${index.uid}/settings`
    const client = new MeiliSearch({ host })
    const strippedHost = trailing ? host.slice(0, -1) : host
    await expect(
      client.index(index.uid).resetSettings()
    ).rejects.toHaveProperty(
      'message',
      `request to ${strippedHost}/${route} failed, reason: connect ECONNREFUSED ${BAD_HOST.replace(
        'http://',
        ''
      )}`
    )
  })
})
