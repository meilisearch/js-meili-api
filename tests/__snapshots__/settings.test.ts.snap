// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on settings Admin key: Get default settings of an index 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Admin key: Get default settings of empty index with primary key 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Admin key: Reset embedders settings  1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Admin key: Reset settings 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Admin key: Reset settings of empty index 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Admin key: Update embedders settings  1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "embedders": {
    "default": {
      "documentTemplate": "{% for field in fields %} {{ field.name }}: {{ field.value }}
{% endfor %}",
      "model": "sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2",
      "source": "huggingFace",
    },
  },
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Admin key: Update searchableAttributes settings on empty index 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "title",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Admin key: Update searchableAttributes settings on empty index with a primary key 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "title",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Admin key: Update settings 1`] = `
{
  "dictionary": [
    "J. K.",
    "J. R. R.",
  ],
  "displayedAttributes": [
    "title",
  ],
  "distinctAttribute": "title",
  "faceting": {
    "maxValuesPerFacet": 50,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [
    "title",
  ],
  "localizedAttributes": null,
  "nonSeparatorTokens": [
    "&sep",
    "/",
    "|",
  ],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byAttribute",
  "rankingRules": [
    "id:asc",
    "typo",
  ],
  "searchCutoffMs": 1000,
  "searchableAttributes": [
    "title",
  ],
  "separatorTokens": [
    "&sep",
    "/",
    "|",
  ],
  "sortableAttributes": [
    "title",
  ],
  "stopWords": [
    "the",
  ],
  "synonyms": {
    "harry": [
      "potter",
    ],
  },
  "typoTolerance": {
    "disableOnAttributes": [
      "comment",
    ],
    "disableOnWords": [
      "prince",
    ],
    "enabled": false,
    "minWordSizeForTypos": {
      "oneTypo": 1,
      "twoTypos": 100,
    },
  },
}
`;

exports[`Test on settings Admin key: Update settings on empty index with primary key 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": "title",
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "title:asc",
    "typo",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [
    "the",
  ],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Admin key: Update settings with all null values 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Master key: Get default settings of an index 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Master key: Get default settings of empty index with primary key 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Master key: Reset embedders settings  1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Master key: Reset settings 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Master key: Reset settings of empty index 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Master key: Update embedders settings  1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "embedders": {
    "default": {
      "documentTemplate": "{% for field in fields %} {{ field.name }}: {{ field.value }}
{% endfor %}",
      "model": "sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2",
      "source": "huggingFace",
    },
  },
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Master key: Update searchableAttributes settings on empty index 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "title",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Master key: Update searchableAttributes settings on empty index with a primary key 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "title",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Master key: Update settings 1`] = `
{
  "dictionary": [
    "J. K.",
    "J. R. R.",
  ],
  "displayedAttributes": [
    "title",
  ],
  "distinctAttribute": "title",
  "faceting": {
    "maxValuesPerFacet": 50,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [
    "title",
  ],
  "localizedAttributes": null,
  "nonSeparatorTokens": [
    "&sep",
    "/",
    "|",
  ],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byAttribute",
  "rankingRules": [
    "id:asc",
    "typo",
  ],
  "searchCutoffMs": 1000,
  "searchableAttributes": [
    "title",
  ],
  "separatorTokens": [
    "&sep",
    "/",
    "|",
  ],
  "sortableAttributes": [
    "title",
  ],
  "stopWords": [
    "the",
  ],
  "synonyms": {
    "harry": [
      "potter",
    ],
  },
  "typoTolerance": {
    "disableOnAttributes": [
      "comment",
    ],
    "disableOnWords": [
      "prince",
    ],
    "enabled": false,
    "minWordSizeForTypos": {
      "oneTypo": 1,
      "twoTypos": 100,
    },
  },
}
`;

exports[`Test on settings Master key: Update settings on empty index with primary key 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": "title",
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "title:asc",
    "typo",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [
    "the",
  ],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;

exports[`Test on settings Master key: Update settings with all null values 1`] = `
{
  "dictionary": [],
  "displayedAttributes": [
    "*",
  ],
  "distinctAttribute": null,
  "faceting": {
    "maxValuesPerFacet": 100,
    "sortFacetValuesBy": {
      "*": "alpha",
    },
  },
  "filterableAttributes": [],
  "localizedAttributes": null,
  "nonSeparatorTokens": [],
  "pagination": {
    "maxTotalHits": 1000,
  },
  "proximityPrecision": "byWord",
  "rankingRules": [
    "words",
    "typo",
    "proximity",
    "attribute",
    "sort",
    "exactness",
  ],
  "searchCutoffMs": null,
  "searchableAttributes": [
    "*",
  ],
  "separatorTokens": [],
  "sortableAttributes": [],
  "stopWords": [],
  "synonyms": {},
  "typoTolerance": {
    "disableOnAttributes": [],
    "disableOnWords": [],
    "enabled": true,
    "minWordSizeForTypos": {
      "oneTypo": 5,
      "twoTypos": 9,
    },
  },
}
`;
